primitives/src/merkle_tree/macros.rs

```diff
@@ -0,0 +1,248 @@
+                    if !elem.eq(element.borrow()) {
+                        return Err(PrimitivesError::ParameterError(
+                            "Element does not match the proof.".to_string(),
+                        ));
+                    }
+                    let proof_leaf_value = H::digest_leaf(pos, elem);
+                    let mut path_values = vec![proof_leaf_value];
+                    traversal_path.iter().zip(proof.proof.iter().skip(1)).fold(
+                        Ok(proof_leaf_value),
+                        |result, (branch, node)| -> Result<T, PrimitivesError> {
+                            match result {
+                                Ok(val) => match node {
+                                    MerkleNode::Branch { value: _, children } => {
+                                        let mut data: Vec<_> =
+                                            children.iter().map(|node| node.value()).collect();
+                                        data[*branch] = val;
+                                        let digest = H::digest(&data);
+                                        path_values.push(digest);
+                                        Ok(digest)
+                                    },
+                                    _ => Err(PrimitivesError::ParameterError(
+                                        "Incompatible proof for this merkle tree".to_string(),
+                                    )),
+                                },
+                                Err(e) => Err(e),
+                            }
+                        },
+                    )?;
+                    self.root.remember_internal::<H, Arity>(
+                        self.height,
+                        &traversal_path,
+                        &path_values,
+                        &proof.proof,
+                    )
+                } else {
+                    Err(PrimitivesError::ParameterError(
+                        "Invalid proof type".to_string(),
+                    ))
+                }
+            }
+        }
+    };
+}
+
+/// Macros for implementing ToTreversalPath for primitive types
+#[macro_export]
+macro_rules! impl_to_traversal_path_primitives {
+    ($t: ty) => {
+        impl<Arity: Unsigned> ToTraversalPath<Arity> for $t {
+            fn to_traversal_path(&self, height: usize) -> Vec<usize> {
+                let mut pos = *self as u64;
+                let mut ret = vec![];
+                for _i in 0..height {
+                    ret.push((pos % (Arity::to_u64())).to_usize().unwrap());
+                    pos /= Arity::to_u64();
+                }
+                ret
+            }
+        }
+    };
+}
+
+/// Macros for implementing ToTreversalPath for BigUint types
+#[macro_export]
+macro_rules! impl_to_traversal_path_biguint {
+    ($t: ty) => {
+        impl<Arity: Unsigned> ToTraversalPath<Arity> for $t {
+            fn to_traversal_path(&self, height: usize) -> Vec<usize> {
```

Comment:
curious, do you want to merge these two into one macro to save code duplication?

impl_to_traversal_path(type, is_primitive) ?